[BEAM-11482] Thrift support for KafkaTableProvider

ccciudatu opened the pull request:
This is a follow-up for https://github.com/apache/beam/pull/13428 to add Thrift support in `KafkaTableProvider`, since we now have all the building blocks for a `BeamKafkaThriftTable`.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

ccciudatu commented:
R: @TheNeuralBit @chrlarsen

ccciudatu commented:
@TheNeuralBit Is installing the `thrift` binary on Jenkins servers an option?
Otherwise, I'll have to fall back to generating thrift manually and add it to source control, the same as @chrlarsen and I did within the `io` module.

ccciudatu commented:
R: @piotr-szuberski

piotr-szuberski commented on CHANGES.md:
```diff
@@ -66,6 +66,7 @@
 * Added Cloud Bigtable Provider extension to Beam SQL ([BEAM-11173](https://issues.apache.org/jira/browse/BEAM-11173), [BEAM-11373](https://issues.apache.org/jira/browse/BEAM-11373))
 * Added a schema provider for thrift data ([BEAM-11338](https://issues.apache.org/jira/browse/BEAM-11338))
 * Added combiner packing pipeline optimization to Dataflow runner. ([BEAM-10641](https://issues.apache.org/jira/browse/BEAM-10641))
+* Added support for thrift in KafkaTableProvider ([BEAM-11482](https://issues.apache.org/jira/browse/BEAM-11482))
```
It would be good to update the external table documentation (create-external-table.md) as well.

ccciudatu replied:
I wasn't aware of that. I'll look it up and update it, thanks for pointing it out.

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+  private final TypeDescriptor<T> typeDescriptor;
+
+  public BeamKafkaThriftTable(
+      @NonNull Schema requiredSchema,
```
I may be wrong but I think that`@NonNull` annotations are quite redundant as it's the default for the checker and only `@Nullable` params are worth annotating. The same below.

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+
+  private static Schema thriftSchema(
+      @NonNull Class<?> thriftClass, @NonNull Schema requiredSchema) {
+    @SuppressWarnings("nullness")
```
Can't it be written without suppressing the nullness? Sometimes checker is awkward and requires some workarounds but I think it should be possible.

ccciudatu replied:
I'll try that.

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+  private static Schema thriftSchema(
+      @NonNull Class<?> thriftClass, @NonNull Schema requiredSchema) {
+    @SuppressWarnings("nullness")
+    final @NonNull Schema thriftSchema =
```
As the rest of `@NonNull` when a param is not `@Nullable` we assume it can't be null

ccciudatu replied:
`SchemaProvider.schemaFor()` returns a `@Nullable` schema, so I was trying to force a "cast" to `@NonNull` here.
Looks like the only elegant way around this is to use `Preconditions.checkArgumentNotNull()` (the beam preconditions, as the guava ones or the standard `Objects.requireNonNull()` won't work).

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+    final @NonNull SchemaProvider schemaProvider = ThriftSchema.provider();
+    return new PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>>() {
+      @Override
+      @SuppressWarnings("nullness")
```
I'd try to get rid of this nullness suppression. Usually if (var != null) {} else { throw NullPointerException("var was null") is unfortunatelly needed to achieve that (that's my experience, maybe there is some cleaner way that I don't know about)

ccciudatu replied:
I'll give it a try.

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+      public PCollection<Row> expand(PCollection<KV<byte[], byte[]>> input) {
+        return input
+            .apply(Values.create())
+            .apply(MapElements.into(typeDescriptor).via(BeamKafkaThriftTable.this::decode))
```
Wouldn't this::decode be sufficient?

piotr-szuberski replied:
Just one more comment.
It would be good to have some function or ParDo declared like it is done for protobuf so we don't have to repeat methods like encode and decode together with ThriftCoder in every table provider that supports thrift.
We could then just import a function returning SimpleFunction/ParDo and use it like `MapElements.via(thriftBytesToBeamRow(thriftClass, protocolFactory))`

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+    try {
+      return thriftCoder.decode(new ByteArrayInputStream(bytes));
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
```
Could you add some message to this exception?

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -0,0 +1,128 @@
+      thriftCoder.encode(thrift, out);
+      return out.toByteArray();
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
```
Please add some message to the exception

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:
```diff
@@ -85,6 +89,33 @@ public BeamSqlTable buildBeamSqlTable(Table table) {
           throw new IllegalArgumentException("Incorrect proto class provided: " + protoClassName);
         }
+      case THRIFT:
+        final String thriftClassName = properties.getString("thriftClass");
```
I'd delegate this part of code to a function. It'll get a bit messy after we add another case to the switch statement.

ccciudatu replied:
Everything else was inlined here, so I thought I'd keep it "consistent".
But you're right, thrift also requires a protocol spec, which makes it even more involved than the proto case, so it probably deserves its own method.

piotr-szuberski replied:
I agree that at least PROTO case should be delegated to a method as well

ccciudatu replied:
I'll extract that one too

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:
```diff
@@ -85,6 +89,33 @@ public BeamSqlTable buildBeamSqlTable(Table table) {
+              | IllegalAccessException
+              | InvocationTargetException
+              | NoSuchMethodException e) {
+            throw new IllegalStateException(e);
```
Please add a message here as well.

piotr-szuberski commented on sdks/java/extensions/sql/src/test/thrift/kafka/message.thrift:
```diff
@@ -0,0 +1,31 @@
+ */
+
+/*
+thrift --gen java:private-members \
```
+1 for this comment - it'll be helpful :)

ccciudatu replied:
I initially tried to generate java code from thrift "on the fly" using the `thrift-gradle-plugin`, but it turns out `thrift` is not installed on the Jenkins servers so the build fails. That's why I had to document how the code is to be generated manually.

piotr-szuberski replied:
Yeah, I read the rest of the comments, that's why it's neat to have this comment here :)

piotr-szuberski commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -183,7 +184,12 @@ private Schema schemaFor(Class<?> targetClass) {
     try {
       final FieldType type = beamType(fieldDescriptor.valueMetaData);
-      return Schema.Field.nullable(fieldDescriptor.fieldName, type);
+      switch (fieldDescriptor.requirementType) {
```
`if {} else {}` or  `return fieldDescriptor.requirementType == REQUIRED ? ... : ...` would be neater as there is only one case.

ccciudatu replied:
Agreed.
My intention here was to leave room for special handling of the `DEFAULT` "field requiredness", as that's somewhere in between, i.e. defined as "opt-in, req-out", so it may not always fall into the "nullable" default case, the same as `OPTIONAL`.
Would it be better if I listed all the options (with some comments, perhaps) and just fall-through to the default (nullable) case?
I can also switch (pun intended) to an if/else or ternary statement, if you think we don't have to make it obvious that each thrift requirement type can be handled separately.

piotr-szuberski replied:
As I said I don't know almost anything about Thrift, so I agree to leave it in the switch statement if it makes more sense.

piotr-szuberski reviewed (changes requested):
Looks good! I've left few comments, mostly cosmetic.

I don't have any knowledge about Thrift though so I'd suggest adding someone who took part in it.

piotr-szuberski commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProviderThriftIT.java:
```diff
@@ -0,0 +1,62 @@
+/*
```
Could you also add a test that will run on SQL precommit as well? You could use the same thrift file as in IT tests.
Example here: https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaTableAvroTest.java

ccciudatu replied:
Sure, I'll add those.

ccciudatu commented:
@piotr-szuberski I pushed a fixup to address all your comments. I'll double check if I missed anything, but please have another look.

piotr-szuberski commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProviderTest.java:
```diff
@@ -64,15 +66,34 @@ public void testBuildBeamSqlAvroTable() {
-    BeamKafkaProtoTable csvTable = (BeamKafkaProtoTable) sqlTable;
-    assertEquals("localhost:9092", csvTable.getBootstrapServers());
-    assertEquals(ImmutableList.of("topic1", "topic2"), csvTable.getTopics());
+    BeamKafkaProtoTable protoTable = (BeamKafkaProtoTable) sqlTable;
```
Good catch! Thanks!

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java:
```diff
@@ -74,55 +92,81 @@ private static Schema thriftSchema(
-      return thriftCoder.decode(new ByteArrayInputStream(bytes));
-    } catch (IOException e) {
-      throw new IllegalStateException(e);
+  private static class InputTransformer<T extends TBase<?, ?>>
```
What I meant was to create some SerializableFunction or SimpleFuntion outside of this package (e.g. in  io/thrift) that would do the encoding/decoding. Right now we'll have to implement encode/decode it in every table provider that supports thrift format. You wouldn't need InputTransformer and OutputTransformer then.

What I have in mind it's the same as is done here:
https://github.com/apache/beam/blob/1466db958ca22222ae8badf1f21a860267d46d31/sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java#L132

And then here:
https://github.com/apache/beam/blob/1466db958ca22222ae8badf1f21a860267d46d31/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java#L83

only for Thrift encoding/decoding

ccciudatu replied:
Got it, makes sense. I tried to solve it locally for Kafka, as I think making this truly reusable/composable requires a larger refactoring that should follow up. But those functions do seem to rather belong in the `ThriftSchema` anyway, so I'll fix this.

ccciudatu replied:
I had a closer look and it seems like the proto `bytesToRow` and `rowToBytes` code has no real reason to be proto specific, as it can work with any `SchemaProvider` and `Coder` pair. I'll try to extract that into `org.apache.beam.sdk.util` and reuse it from both proto and thrift (and perhaps others, if I find more like this). This way I'll also add support for shuffled schemas in the thrift implementation for free.

ccciudatu replied:
Just pushed another fixup for this. It's pretty big and I'll try to draw your attention to the potentially dangerous/incompatible changes.

ccciudatu replied:
I can revert the change breaking change in proto (i.e. `Class<?>` -> `Class<? extends Message>`) and use some ugly casts internally, if this is a concern.

ccciudatu replied:
Nevermind, I just noticed the comments which read:

    // Other modules are not allowed to use non-vendored Message class

So I'll revert that change.

piotr-szuberski reviewed (changes requested):
Thanks! I've left only one comment.

ccciudatu commented:
Squashed it all in a single commit and rebased on top of master.

ccciudatu commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java:
```diff
@@ -32,10 +34,10 @@
 
   public BeamKafkaProtoTable(
-      Schema messageSchema, String bootstrapServers, List<String> topics, Class<?> protoClass) {
+      Schema messageSchema, String bootstrapServers, List<String> topics, Class<? extends Message> protoClass) {
```
This change breaks any clients that used to pass a mere `Class<?>` here!
Is this acceptable?

ccciudatu replied:
reverted

ccciudatu commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java:
```diff
@@ -95,9 +97,9 @@ private static Schema inferAndVerifySchema(Class<?> protoClass, Schema messageSc
+    private final Class<? extends Message> clazz;
 
-    public ProtoRecorderEncoder(Class<?> clazz) {
+    public ProtoRecorderEncoder(Class<? extends Message> clazz) {
```
Is this acceptable?

ccciudatu replied:
reverted

ccciudatu commented on sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java:
```diff
@@ -129,68 +129,29 @@ public SchemaUserTypeCreator schemaTypeCreator(Class<?> targetClass, Schema sche
     "unchecked"
   })
-  public static <T> SimpleFunction<byte[], Row> getProtoBytesToRowFn(Class<T> clazz) {
+  public static <T extends Message> SimpleFunction<byte[], Row> getProtoBytesToRowFn(
```
This again breaks compatibility with `Class<?>`.

ccciudatu replied:
reverted

ccciudatu commented on sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java:
```diff
@@ -129,68 +129,29 @@ public SchemaUserTypeCreator schemaTypeCreator(Class<?> targetClass, Schema sche
     "unchecked"
   })
-  public static <T> SimpleFunction<Row, byte[]> getRowToProtoBytesFn(Class<T> clazz) {
+  public static <T extends Message> SimpleFunction<Row, byte[]> getRowToProtoBytesFn(
```
`Class<?>` won't work anymore

ccciudatu replied:
reverted

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftCoder.java:
```diff
@@ -72,15 +71,13 @@ protected ThriftCoder(Class<T> type, TProtocolFactory protocolFactory) {
   public void encode(T value, OutputStream outStream) throws CoderException, IOException {
-    ByteArrayOutputStream baos = new ByteArrayOutputStream();
-    TProtocol protocol = protocolFactory.getProtocol(new TIOStreamTransport(baos));
+    TProtocol protocol = protocolFactory.getProtocol(new TIOStreamTransport(outStream));
```
@chrlarsen Just trying to make sure I haven't missed anything: was there a reason why we're currently writing to a temporary in-memory stream and then copy that to the actual stream? i.e. do you see any potential pitfalls with the optimization that I just made?

chrlarsen replied:
LGTM. I think this was an artifact from earlier development.

ccciudatu commented on sdks/java/build-tools/src/main/resources/beam/suppressions.xml:
```diff
@@ -94,6 +94,7 @@
   <!-- Non-vendored gRPC/protobuf imports are allowed for files that depend on libraries that expose gRPC/protobuf in its public API -->
   <suppress id="ForbidNonVendoredGrpcProtobuf" files=".*sdk.*extensions.*protobuf.*" />
   <suppress id="ForbidNonVendoredGrpcProtobuf" files=".*sdk.*extensions.*ml.*" />
+  <suppress id="ForbidNonVendoredGrpcProtobuf" files=".*sdk.*extensions.*sql.*" />
```
Is this ok?

ccciudatu replied:
reverted

ccciudatu commented on sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java:
```diff
@@ -130,28 +129,12 @@ public SchemaUserTypeCreator schemaTypeCreator(Class<?> targetClass, Schema sche
+    return new SqlRows.BytesToRowFn<>(
+        new ProtoMessageSchema(),
+        TypeDescriptor.of(protoClass),
+        bytes -> protoCoder.getParser().parseFrom(bytes));
```
this can't be a method reference (`protoCoder.getParser()::parseFrom`) because the protobuf `Parser` is not `Serializable`.

ccciudatu commented:
I squashed (once again) all the changes in a single commit and added support for field reordering in `ThriftSchema`, just to add a test case for the shuffled schemas, the same as proto has.

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -243,37 +253,65 @@ private FieldValueTypeInformation fieldValueTypeInfo(Class<?> type, String field
-          actualValue = set;
+          actualValue =
+              StreamSupport.stream(((Iterable<?>) value).spliterator(), false)
+                  .collect(Collectors.toSet());
```
This wasn't possible before because of the "nullness" checks, but now it seems to work so I cleaned it up.

piotr-szuberski reviewed (approved):
LGTM  👍

piotr-szuberski commented:
Run SQL PostCommit

ccciudatu commented:
@piotr-szuberski @TheNeuralBit @iemejia any chance this could end up in the `2.27.0` release or should I revisit the update to `CHANGES.md`?

TheNeuralBit commented on CHANGES.md:
```diff
@@ -66,6 +66,7 @@
 * Added Cloud Bigtable Provider extension to Beam SQL ([BEAM-11173](https://issues.apache.org/jira/browse/BEAM-11173), [BEAM-11373](https://issues.apache.org/jira/browse/BEAM-11373))
 * Added a schema provider for thrift data ([BEAM-11338](https://issues.apache.org/jira/browse/BEAM-11338))
 * Added combiner packing pipeline optimization to Dataflow runner. ([BEAM-10641](https://issues.apache.org/jira/browse/BEAM-10641))
+* Added support for thrift in KafkaTableProvider ([BEAM-11482](https://issues.apache.org/jira/browse/BEAM-11482))
```
Unfortunately this will be in 2.28.0 since 2.27.0 was cut last week. Could you merge master (it should have the 2.28 section), and move this there?

ccciudatu replied:
There's doesn't seem to be a section for `2.28.0` yet.
I was hoping we could get rid of all our custom thrift extensions in our project by just upgrading to `2.27.0`, but I wasn't fast enough. :)
Should I create the `2.28.0` section myself and add move this bullet over there?

TheNeuralBit replied:
Ah yes please go ahead and create the section. Sorry I thought this happened as part of the release cut.

TheNeuralBit replied:
https://github.com/apache/beam/pull/13610 just added the 2.28.0 section so you can merge or rebase to get it

ccciudatu replied:
done

TheNeuralBit reviewed (commented):
A few more requests. This is looking good overall

TheNeuralBit commented:
> any chance this could end up in the 2.27.0 release or should I revisit the update to CHANGES.md?

The branch has already been cut so we'd need to cherrypick it to get it into 2.27.0. Usually that's reserved for bugfixes, but it's up to the discretion of the release manager

ccciudatu commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProviderThriftIT.java:
```diff
@@ -0,0 +1,63 @@
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TProtocolFactory;
+
+public class KafkaTableProviderThriftIT extends KafkaTableProviderIT {
```
done

ccciudatu commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/SqlRows.java:
```diff
@@ -0,0 +1,132 @@
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+public final class SqlRows {
```
done

ccciudatu replied:
It's just a tiny step towards whatever abstractions we want to end up with, as it doesn't even cover Avro yet.
I'm still trying to wrap my head around how would a proper data format abstraction look like in Beam.
I mean, instead of `FileIO`, `AvroIO`, `KafkaIO`, `ThriftIO` plus Kafka tables for each data format, what if we took the format out of the IO as an entirely different concern and provide the ability to mix those freely. If IOs are just `[File, Kafka, PubSub, Socket, whatever]`, and data formats are `[Avro, Protobuf, CSV, Thrift, whatever]`, we can (in theory) allow any combination like PubSub+Thrift or File+Proto with no extra cost (except for incompatible candidates, like Kafka+parquet, perhaps).

Do you think this is worth considering (for 3.0, of course)? I am aware that such an abstraction is really difficult to get right (if at all feasible), but I still think it may be worth trying, as we should at least end up reducing the number of data_source+data_format combinations that are now modelled as new datasource types.

TheNeuralBit replied:
Yes this sounds ideal. As far as 3.0 - I think it would make sense to start working on it in 2.x versions as an `@Experimental` API. We shouldn't remove the existing IOs prior to a major version change, but it would be fine to start work on an alternative. This would be similar to (and related to) the `SchemaIOProvider` concept we had a Google intern work on last summer:
- https://s.apache.org/schemaio-development-guide
- http://s.apache.org/beam-schema-io

Would you have any interest in working on an effort like this?

CC: @robinyqiu a design like this would be relevant for your work on IOs

ccciudatu replied:
I'd be more than happy to get involved. Most likely, I'll need to join your slack channel, and for that I need an apache.org email, apparently. I'll see how I can sort that out...

TheNeuralBit replied:
It looks like you can get an invite at https://s.apache.org/slack-invite (more info here: https://infra.apache.org/slack.html)

ccciudatu replied:
Sounds easy. Not sure who's supposed to invite me, but I'll ask around. If you can help with that, I'd be grateful.

TheNeuralBit replied:
Sure I can invite you as a guest. I think https://s.apache.org/slack-invite is _supposed_ to be self-service but the page indicates it breaks it often. Can you send me a message at bhulette@apache.org so I know what email to use?

TheNeuralBit commented:
Run SQL Postcommit

TheNeuralBit reviewed (approved):
LGTM, thank you @ccciudatu!

TheNeuralBit commented:
@chrlarsen would you have time to take a look? Specifically it would be good to check over the changes to `ThriftSchema`

chrlarsen commented:
@TheNeuralBit definitely I'll have a look.

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -90,17 +95,17 @@
  *   <li>All non-union types have a corresponding java field with the same name for every field in
  *       the original thrift source file.
- *   <li>The underlying {@link FieldMetaData#getStructMetaDataMap(Class) metadata maps} are {@link
- *       java.util.EnumMap enum maps}, so the natural order of the field keys is preserved.
```
This is still the case with thrift metadata, but it's no longer a "strong assumption" for this class, as now the field order is infered from the schema, allowing for field reordering.

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -90,17 +95,17 @@
+ * TBase#isSet(TFieldIdEnum) isSet} flag may not be preserved upon converting a thrift object to a
+ * beam row and back. On encoding, we extract all thrift values, no matter if the fields are set or
+ * not. On decoding, we set all non-{@code null} beam row values to the corresponding thrift fields,
+ * leaving the rest unset.
```
I changed the set-ALL-fields policy to set non-null only, as I think this is safer (no NPE if primitive thrift fields are null in the beam row), easier to reason about and more natural for thrift clients (who are used to check if fields are set before using them).

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -286,8 +324,6 @@ T restoreThriftObject(Class<?> targetClass, Object[] params) {
       }
       thrift.setFieldValue(field, actualValue);
-    } else if (!TUnion.class.isInstance(thrift)) {
-      thrift.setFieldValue(field, value); // nullness checks don't allow setting null here
```
This seemed like a good idea at the time, as it prevented null values in the beam row from ending up as fields with default values in thrift (in case defaults were provided in the thrift descriptor), so it seemed closer to a "symmetrical" transformation.
However, leaving the fields unset when they're null in the source row is both safer and more natural.

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -243,37 +253,65 @@ private FieldValueTypeInformation fieldValueTypeInfo(Class<?> type, String field
+  @SuppressWarnings("nullness")
   private <FieldT extends Enum<FieldT> & TFieldIdEnum, T extends TBase<T, FieldT>>
-      T restoreThriftObject(Class<?> targetClass, Object[] params) {
+      Map<FieldT, FieldMetaData> schemaFieldDescriptors(Class<?> targetClass, Schema schema) {
```
This allows using the order of fields in the incoming schema, instead of the original order in the thrift descriptor, so this will still work if fields are reordered by clients (the shuffled schema table test validates this).

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -183,7 +188,12 @@ private Schema schemaFor(Class<?> targetClass) {
+          return Schema.Field.of(fieldDescriptor.fieldName, type);
+        default:
+          return Schema.Field.nullable(fieldDescriptor.fieldName, type);
+      }
```
Fields used to always be nullable for thrift schemas and the table integration test requires non-nullable fields. So I just added what was actually missing, i.e. marking `required` as non-nullable in beam. `default` is still nullable, the same as `optional`, for safety reasons -- as it's defined as "opt-in, req-out".

ccciudatu commented on sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftSchema.java:
```diff
@@ -243,37 +253,65 @@ private FieldValueTypeInformation fieldValueTypeInfo(Class<?> type, String field
+  }
+
+  private <FieldT extends Enum<FieldT> & TFieldIdEnum, T extends TBase<T, FieldT>>
+      SchemaUserTypeCreator thriftMapper(Class<?> targetClass, Schema schema) {
```
this method only exists for "pinning" the generics between the two invocations

chrlarsen reviewed (approved):
LGTM. The `ThriftSchema` changes look to be in line with what the Thrift IDL specifies, thanks @ccciudatu!

ccciudatu commented:
resolved merge conflicts for `CHANGES.md` and added the following cosmetics:
- simplified generics, as the `extends Enum<..>` constraint is no longer needed
- used the defined overload for proto mockTable in `KafkaTableProviderTest`
- listed the `DEFAULT` and `OPTIONAL` "requiredness" types explicitly, falling through to default -- for self-documenting purposes and to eliminate the compiler warning for too few cases (@piotr-szuberski, I hope you're ok with this)

TheNeuralBit commented:
Run SQL PostCommit

aromanenko-dev commented:
@ccciudatu Thanks for contribution! 

For the future PRs, please, don't use your local master (or main) branch, create a dedicated feature branch for changes.

Perhaps, it might help:
https://cwiki.apache.org/confluence/display/BEAM/Git+Tips

TheNeuralBit commented:
@aromanenko-dev whoops sorry about that, I didn't notice this PR was opened from a master branch. Just curious, did it cause any problems in this case?

aromanenko-dev commented:
@TheNeuralBit Afaict, it should not cause big problems after it was merged (only it won't be possible to delete a "feature" branch in fork and it will rewrite a commit history I guess). Though, until merge, all pushes to fork master branch will appear in the PR which likely won't be acceptable.
